Declare record type for flambda function parameters

chambart opened the pull request:
This will simplify two future patches adding informations to functions parameters

* Per parameter inline annotations
* Function type propagation (in particular for unboxing)

chambart committed:
Declare a new record type for Flambda function parameters

mshinwell commented on middle_end/flambda_utils.ml:
```diff
@@ -79,6 +79,8 @@ let compare_const (c1 : Flambda.const) (c2 : Flambda.const) =
   | _, Char _ -> 1
 
+[@@@ocaml.warning "+9"]
+
```
I'm not sure how this attribute relates to the only change in the middle of the section it applies to (just `Variable.equal` -> `Parameter.equal`)

chambart replied:
Sorry, this was added for Parameter.equal which was moved to another file

mshinwell commented on middle_end/freshening.ml:
```diff
@@ -137,10 +137,11 @@ let add_variable t id =
+  List.fold_right (fun (id:Parameter.t) (ids, t) ->
+      let id', t = active_add_variable t id.var in
+      let param : Parameter.t = { var = id' } in
+      param :: ids, t) params ([], t)
```
Let's put `params ([], t)` on a new line (indented by 2 not 4); I think it's more legible that way.

mshinwell commented on middle_end/inline_and_simplify_aux.ml:
```diff
@@ -543,7 +543,7 @@ let prepare_to_simplify_set_of_closures ~env
           | None -> true
           | Some function_decl ->
-            Variable.Set.mem param (Variable.Set.of_list function_decl.params)
+            Variable.Set.mem param (Parameter.var_set function_decl.params)
```
How about adding `Parameter.Set.mem_var : t -> Variable.t -> bool` ?

chambart replied:
I will do that.

mshinwell commented on middle_end/inlining_transforms.ml:
```diff
@@ -82,6 +82,7 @@ let set_inline_attribute_on_all_apply body inline specialise =
 let copy_of_function's_body_with_freshened_params env
       ~(function_decl : Flambda.function_declaration) =
   let params = function_decl.params in
+  let params_var = Parameter.vars params in
```
`param_vars` not `params_var` is probably better.

mshinwell commented on middle_end/inlining_transforms.ml:
```diff
@@ -90,13 +91,20 @@ let copy_of_function's_body_with_freshened_params env
-    let subst = Variable.Map.of_list (List.combine params freshened_params) in
+    let freshened_params =
+      List.map (fun (p:Parameter.t) : Parameter.t ->
+        { var = Variable.rename p.var })
```
Indent line 101 by two more please, for clarity

mshinwell replied:
How about adding `Parameter.rename`?

mshinwell commented on middle_end/parameter.mli:
```diff
@@ -0,0 +1,27 @@
+
+type t = {
+  var : Variable.t;
+}
```
I think we should make this abstract with a creation function.  It will probably be better in the long term, and I don't think it's much of a change to this patch.

chambart replied:
It wasn't too bothersome and caught some misspropagated parameters

mshinwell commented on middle_end/parameter.mli:
```diff
@@ -0,0 +1,27 @@
+
+val var : t -> Variable.t
+val vars : t list -> Variable.t list
+val var_set : t list -> Variable.Set.t
```
Calls to `Parameter.vars` look a bit peculiar.  How about moving `Parameter.vars` to `Parameter.List.vars` and the same for `var_set`?

mshinwell commented on middle_end/parameter.mli:
```diff
@@ -0,0 +1,27 @@
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
```
Please add a short comment with a description of the module.

mshinwell commented on middle_end/remove_unused_arguments.ml:
```diff
@@ -43,13 +44,20 @@ let remove_params unused (fun_decl: Flambda.function_declaration) =
   let renamed = rename_var var in
+  let rename_param (param:Parameter.t) : Parameter.t =
+    { var = rename_var param.var }
+  in
```
As above, `Parameter.rename`.

mshinwell commented on middle_end/parameter.mli:
```diff
@@ -0,0 +1,27 @@
+}
+
+val var : t -> Variable.t
+val vars : t list -> Variable.t list
```
We should try to document functions in the interface files.  `vars` in particular should probably have a comment that says that it preserves the order, since that's important.

mshinwell reviewed (changes requested):
For ease of review, please don't squash subsequent changes onto the existing changesets.

chambart committed:
Remove spurious warning annotation

chambart committed:
Add Parameter.rename

chambart committed:
Code style

chambart committed:
Abstract Parameter.t

chambart committed:
Rename Parameter.vars into Parameter.List.vars

mshinwell commented on middle_end/parameter.mli:
```diff
@@ -16,9 +16,13 @@
-  var : Variable.t;
-}
+(** [Parameter.t] carries a unique [Variable.t] used as function parameter.
+    It can also carries annotation on the usage of the variable *)
```
"It can also carry annotations about the usage of the variable."

mshinwell commented:
@chambart I made one comment about a comment; please rebase and merge after fixing that.

mshinwell commented:
@chambart Ping

chambart committed:
Declare module types of Identifiable.Make submodules

This allows to add fields on some instantiations without copying the signature

chambart committed:
Rename Parameter.var_set to Parameter.Set.vars

chambart committed:
Rename var_set uses

chambart committed:
.depend

chambart commented:
I've rebased, fixed the comment, and renamed var_set to Set.vars.
To do this last change I had to move things a bit in Identifiable. In particular, submodules of Identifiable.Make needed to have a module type name that can be refered to.

mshinwell commented:
Something went wrong with the no-change-entry-needed label and the Travis check for such, but the rest of the build is fine, so merging.
